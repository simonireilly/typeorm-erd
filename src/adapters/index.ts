import { DataSource } from "typeorm";
import { MermaidErd } from "./mermaid";
import { PlantUMLErd } from "./plantuml";

const BUILDERS = {
  mermaid: MermaidErd,
  plantuml: PlantUMLErd,
};

type TBuildTypes = keyof typeof BUILDERS;

class ERDBuilder {
  private instance: MermaidErd | PlantUMLErd;

  constructor(public instanceType: TBuildTypes, dataSource: DataSource) {
    const adapter = BUILDERS[instanceType];
    this.instance = new adapter(dataSource);
  }

  public async initialize() {
    return this.instance.initialize();
  }

  public render(): string {
    return this.instance.render();
  }
}

export { PlantUMLErd, MermaidErd, ERDBuilder };
