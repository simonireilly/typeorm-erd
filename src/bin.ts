#!/usr/bin/env node

import { ERDBuilder } from "./index";
import { resolve } from "path";

const main = async (path: string = "./examples/basic/src/data-source") => {
  const filePath = resolve(path);
  console.log("CWD Stats:", { cwd: process.cwd(), filePath });

  const moduleA = await import(filePath);
  console.info("Module", moduleA);
  const erd = new ERDBuilder("mermaid", moduleA);
  await erd.initialize();
  const erdText = await erd.render();

  console.info(erdText);
};

main();
