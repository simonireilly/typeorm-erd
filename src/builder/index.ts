import { DataSource, EntityMetadata, EntitySchema } from "typeorm";
import { ConnectionMetadataBuilder } from "typeorm/connection/ConnectionMetadataBuilder";
import { RelationType } from "typeorm/metadata/types/RelationTypes";

export type BuilderRelations = Record<
  string,
  {
    entityRelations: RelationData[];
  }
>;

interface RelationData {
  relationType: RelationType;
  propertyPath: string;
  nullable: boolean;
  isOwning: boolean;
  inverseSidePropertyPath: string;
  source: string;
  target: string;
}

const entityMetaData = async (connection: DataSource) => {
  const connectionMetadataBuilder = new ConnectionMetadataBuilder(connection);

  const { entities } = connection.options;
  const TEntities = entities as (Function | EntitySchema<any> | string)[];

  let entityMetadata;

  if (entities) {
    entityMetadata = await connectionMetadataBuilder.buildEntityMetadatas(
      TEntities
    );
  } else {
    throw Error("No entities found on connection");
  }

  return entityMetadata;
};

/**
 * Return the relation data for each table
 *
 */
const relations = (meta: EntityMetadata[]): BuilderRelations => {
  const relationData = meta.reduce<BuilderRelations>((acc, entity) => {
    const entityRelations = entity.ownRelations.map(
      ({
        relationType,
        inverseEntityMetadata,
        propertyPath,
        inverseSidePropertyPath,
        isOwning,
      }) => {
        return {
          relationType,
          propertyPath,
          isOwning,
          nullable:
            entity.columns.find((c) => c.propertyName === propertyPath)
              ?.isNullable || false,
          inverseSidePropertyPath,
          source: entity.tableName,
          target: inverseEntityMetadata.tableName,
        };
      }
    );

    return {
      ...acc,
      [entity.tableName]: {
        entityRelations,
      },
    };
  }, {});

  return relationData;
};

export const builders = {
  entityMetaData,
  relations,
};
