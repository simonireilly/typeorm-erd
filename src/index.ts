import { AppDataSource } from "../examples/basic/src/data-source";
import { builders } from "./builder";

const meta = builders.entityMetaData(AppDataSource);
