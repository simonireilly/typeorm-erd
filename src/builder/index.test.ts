import { AppDataSource } from "../../examples/basic/src/data-source";
import { builders } from "./";

describe("building test data", () => {
  it("returns the meta data", () => {
    const meta = builders.entityMetaData(AppDataSource);

    expect(meta).toMatchSnapshot();
  });
});
