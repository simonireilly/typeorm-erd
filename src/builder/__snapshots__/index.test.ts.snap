// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`building test data returns the meta data 1`] = `Promise {}`;
